perf(matrices): Use DomainMatrix to speed up charpoly

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Follows gh-25369

See also gh-25410 for background discussion.

#### Brief description of what is fixed or changed

Change `Matrix.charpoly()` to call `DomainMatrix.charpoly()`.

Also there are some improvements to the `DomainMatrix.charpoly()` method:

- A sparse implementation of the Berkowitz algorithm is added to DomainMatrix (`sdm_berk`).
- Clearing denominators is now used in `DomainMatrix.charpoly()` to compute the characteristic polynomial over a field more efficiently.
- Now `charpoly` automatically selects the best implementation (sparse or dense).

#### Other comments

I will show a few timings below. I have done detailed timings of different algorithms like in gh-25410 but I am not going to present those because the conclusions are extremely simple:

- The sparse implementation of Berkowitz is more efficient if fewer than 50% of the entries in the matrix are nonzero. Otherwise the dense implementation is fastest.
- The use of `charpoly_factor_blocks` (strongly connected components) brings a big speedup for large very sparse matrices as well as smaller structured matrices with complicated entries and never slows anything down significantly.
- Clearing denominators is always beneficial if it can be done.
- Having `Matrix.charpoly()` use `DomainMatrix.charpoly()` is as far as I can tell always faster.

From the timings I've looked at none of these conclusions depends particularly on the type of the expressions, size of the matrix, bit size etc so the choice of algorithm is very simple:

- Use sparse for `<50%` density
- Always clear denominators
- Matrix will be faster in all cases if it uses DomainMatrix.

The output of charpoly is not expected to be changed in any way because it is always converted to Poly anyway. This also means that there is no need to worry about DomainMatrix doing things like expanding powers etc since that would always have happened when converting to `Poly`.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
   * Matrix.charpoly() now uses DomainMatrix to compute the characteristic polynomial of a matrix making it a lot faster. For dense integer matrices charpoly is now about 10x faster. For more complicated symbolic matrices (e.g. with polynomial entries) it is much, much faster so some calculations that would have effectively taken forever can now be completed in reasonable time.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
feat(matrices): add sparse Berkowitz algorithm

oscarbenjamin committed:
perf(matrices): optimise DomainMatrix.charpoly

oscarbenjamin committed:
doc(matrices): fix DomainMatrix docstrings

oscarbenjamin committed:
perf(matrices): use DM for Matrix.charpoly()

oscarbenjamin commented:
Here are timings for integer matrices with the PR:
```python
In [19]: from time import time

In [20]: def time_ms(f):
    ...:     tstart = time()
    ...:     f()
    ...:     return (1e3 * (time() - tstart))
    ...: 

In [21]: for n in [1,2,5,10,20,50,100]:
    ...:     M = randMatrix(n)
    ...:     print(f'{n} x {n} dense integer matrix: {time_ms(M.charpoly):.0f} ms')
    ...: 
1 x 1 dense integer matrix: 1 ms
2 x 2 dense integer matrix: 1 ms
5 x 5 dense integer matrix: 4 ms
10 x 10 dense integer matrix: 9 ms
20 x 20 dense integer matrix: 19 ms
50 x 50 dense integer matrix: 272 ms
100 x 100 dense integer matrix: 4445 ms
```
With master that is:
```python
1 x 1 dense integer matrix: 126 ms
2 x 2 dense integer matrix: 2 ms
5 x 5 dense integer matrix: 4 ms
10 x 10 dense integer matrix: 38 ms
20 x 20 dense integer matrix: 122 ms
50 x 50 dense integer matrix: 1978 ms
100 x 100 dense integer matrix: 25710 ms
```
So the PR is about 5x faster than master for integer matrices. I think as you scale up it works out more like 10x faster but is not asymptotically different for dense integer matrices.

For rational matrices with master we have:
```
In [5]: for n in [1,2,5,10,20,50,100]:
   ...:     M1 = randMatrix(n)
   ...:     M2 = randMatrix(n, min=1)
   ...:     M = Matrix(n, n, lambda i, j: M1[i,j] / M2[i,j])
   ...:     print(f'{n} x {n} dense rational matrix: {time_ms(M.charpoly):.0f} ms')
   ...: 
1 x 1 dense rational matrix: 2 ms
2 x 2 dense rational matrix: 3 ms
5 x 5 dense rational matrix: 14 ms
10 x 10 dense rational matrix: 32 ms
20 x 20 dense rational matrix: 329 ms
50 x 50 dense rational matrix: 41441 ms
^C^C
KeyboardInterrupt
```
With the PR this is a lot faster (asmyptotically faster):
```python
1 x 1 dense rational matrix (PR): 2 ms
2 x 2 dense rational matrix (PR): 2 ms
5 x 5 dense rational matrix (PR): 5 ms
10 x 10 dense rational matrix (PR): 5 ms
20 x 20 dense rational matrix (PR): 17 ms
50 x 50 dense rational matrix (PR): 537 ms
100 x 100 dense rational matrix (PR): 14584 ms
```
At 50x50 it is already a 100x difference and the difference increases as n increases.

For symbolic matrices the differences are more extreme. We'll try matrices like this one:
```python
In [7]: M = randMatrix(5)*x + randMatrix(5)

In [8]: M
Out[8]: 
⎡92⋅x + 29  77⋅x + 27  43⋅x + 13  51⋅x + 90  51⋅x + 99⎤
⎢                                                     ⎥
⎢81⋅x + 51  15⋅x + 72  22⋅x + 42  58⋅x + 20  55⋅x + 16⎥
⎢                                                     ⎥
⎢20⋅x + 57   8⋅x + 3   59⋅x + 31  59⋅x + 39  50⋅x + 75⎥
⎢                                                     ⎥
⎢4⋅x + 88   92⋅x + 86  26⋅x + 61  33⋅x + 22  49⋅x + 29⎥
⎢                                                     ⎥
⎣33⋅x + 36  98⋅x + 21  72⋅x + 16  45⋅x + 39  72⋅x + 77⎦
```
With master that gives:
```python
In [5]: for n in [1,2,3,4,5,6,7,8,9,10]:
   ...:     M = randMatrix(n) + randMatrix(n)*x
   ...:     print(f'{n} x {n} dense symbolic matrix (master): {time_ms(M.charpoly):.0f} ms')
   ...: 
1 x 1 dense symbolic matrix (master): 276 ms
2 x 2 dense symbolic matrix (master): 104 ms
3 x 3 dense symbolic matrix (master): 345 ms
4 x 4 dense symbolic matrix (master): 1373 ms
5 x 5 dense symbolic matrix (master): 9288 ms
6 x 6 dense symbolic matrix (master): 80805 ms
^C^C
KeyboardInterrupt
```
So it is over 1 minute for a 6x6 matrix. The asymptotic scaling looks almost like `10^n` but I wouldn't be surprised if it is worse than exponential.

With the PR the same symbolic matrices can be handled in reasonable time:
```python
In [4]: for n in [1,2,3,4,5,6,7,8,9,10]:
   ...:     M = randMatrix(n) + randMatrix(n)*x
   ...:     print(f'{n} x {n} dense symbolic matrix (PR): {time_ms(M.charpoly):.0f} ms')
   ...: 
1 x 1 dense symbolic matrix (PR): 58 ms
2 x 2 dense symbolic matrix (PR): 66 ms
3 x 3 dense symbolic matrix (PR): 82 ms
4 x 4 dense symbolic matrix (PR): 121 ms
5 x 5 dense symbolic matrix (PR): 166 ms
6 x 6 dense symbolic matrix (PR): 227 ms
7 x 7 dense symbolic matrix (PR): 465 ms
8 x 8 dense symbolic matrix (PR): 466 ms
9 x 9 dense symbolic matrix (PR): 507 ms
10 x 10 dense symbolic matrix (PR): 639 ms
```
At 6x6 this is about 350x faster but the ratio will increase as n increases. For the 80 seconds that it takes master to compute the charpoly of a 6x6 matrix we could instead compute the charpoly of a 50x50 matrix with the PR so that is a significant increase in problem size.

If the domain ends up being `EX` then this might not be faster than master e.g. this happens in about the same time for both master and the PR:
```python
In [18]: M = randMatrix(5)*sin(x) + randMatrix(5)*cos(x)

In [19]: %time ok = M.charpoly()
CPU times: user 13 s, sys: 0 ns, total: 13 s
Wall time: 13 s
```
We could contemplate not using `DomainMatrix` for these domains since it does not necessarily offer any improvement and potentially the existing routines are better optimised for the EX case.

Note that this particular example would be handled much more efficiently if we had a domain that could represent sin and cos or also if we just make DomainMatrix a bit greedier about what it allows for a domain. In this case since we don't need zero detection it is fine to treat sin and cos as algebraically independent (although much more efficient to handle them as algebraically dependent)/ For example with the PR we could do either of these:
```python
In [1]: M = randMatrix(5)*sin(x) + randMatrix(5)*cos(x)

In [2]: %time ok = M.to_DM(ZZ[sin(x),cos(x)]).charpoly()
CPU times: user 25.7 ms, sys: 210 µs, total: 25.9 ms
Wall time: 25.1 ms

In [3]: %time ok = M.to_DM(EX).charpoly()
CPU times: user 2.19 s, sys: 356 µs, total: 2.19 s
Wall time: 2.19 s
```
Currently this uses `EXRAW` rather than `EX` which is similar to the behaviour with master. In this case though `EX` is actually more efficient because it is better to expand along the way rather than expand at the end.

This PR does not use extension=True to handle algebraic numbers fields like `Q(sqrt(2))` but that is something that could be added (with care to avoid `primitive_element` being slow):
```python
In [4]: M = randMatrix(5) + randMatrix(5)*sqrt(2)

In [5]: M
Out[5]: 
⎡13 + 42⋅√2  39 + 95⋅√2  4⋅√2 + 42   54 + 90⋅√2  62⋅√2 + 94⎤
⎢                                                          ⎥
⎢62 + 63⋅√2   √2 + 59    30⋅√2 + 58  5 + 58⋅√2   18 + 28⋅√2⎥
⎢                                                          ⎥
⎢34⋅√2 + 93  17 + 68⋅√2   √2 + 80    26 + 37⋅√2  39 + 80⋅√2⎥
⎢                                                          ⎥
⎢59 + 54⋅√2   7 + 5⋅√2   38 + 51⋅√2  30⋅√2 + 76  19 + 73⋅√2⎥
⎢                                                          ⎥
⎣63 + 92⋅√2  14⋅√2 + 54  54 + 58⋅√2  45⋅√2 + 70  17⋅√2 + 28⎦

In [6]: %time ok = M.charpoly()
CPU times: user 29.3 s, sys: 6.55 ms, total: 29.3 s
Wall time: 29.3 s

In [7]: %time ok = M.to_DM(EX).charpoly()
CPU times: user 1.1 s, sys: 0 ns, total: 1.1 s
Wall time: 1.1 s

In [8]: %time ok = M.to_DM(extension=True).charpoly()
CPU times: user 43.2 ms, sys: 4.07 ms, total: 47.3 ms
Wall time: 46 ms
```
With master that is:
```python
In [1]: M = randMatrix(5) + randMatrix(5)*sqrt(2)

In [2]: %time ok = M.charpoly()
CPU times: user 31.8 s, sys: 18.1 ms, total: 31.8 s
Wall time: 31.8 s
```
So again when we fall back on `EXRAW` the times are about the same as master but we could get much better timings if we could always choose the best domain. Both these last two examples actually suggest that it would be better to fall back on `EX` rather than `EXRAW` at least for `charpoly` so maybe that should be changed...

With Gaussian integers or rationals the domain will be picked up automatically because `construct_domain` handles those automatically. Master is very slow in that case:
```python
In [6]: for n in [1,2,3,4,5,6,7,8,9,10]:
   ...:     M = randMatrix(n) + randMatrix(n)*I
   ...:     print(f'{n} x {n} dense ZZ_I matrix (master): {time_ms(M.charpoly):.0f} ms')
   ...: 
1 x 1 dense ZZ_I matrix (master): 43 ms
2 x 2 dense ZZ_I matrix (master): 97 ms
3 x 3 dense ZZ_I matrix (master): 398 ms
4 x 4 dense ZZ_I matrix (master): 2282 ms
5 x 5 dense ZZ_I matrix (master): 17260 ms
^C^C
KeyboardInterrupt
```
With the PR it is much faster:
```python
In [4]: for n in [1,2,3,4,5,6,7,8,9,10]:
   ...:     M = randMatrix(n) + randMatrix(n)*I
   ...:     print(f'{n} x {n} dense ZZ_I matrix (PR): {time_ms(M.charpoly):.0f} ms')
   ...: 
1 x 1 dense ZZ_I matrix (PR): 37 ms
2 x 2 dense ZZ_I matrix (PR): 32 ms
3 x 3 dense ZZ_I matrix (PR): 45 ms
4 x 4 dense ZZ_I matrix (PR): 60 ms
5 x 5 dense ZZ_I matrix (PR): 78 ms
6 x 6 dense ZZ_I matrix (PR): 89 ms
7 x 7 dense ZZ_I matrix (PR): 100 ms
8 x 8 dense ZZ_I matrix (PR): 113 ms
9 x 9 dense ZZ_I matrix (PR): 138 ms
10 x 10 dense ZZ_I matrix (PR): 150 ms
```
This is also a case that can be improved by speeding up `ZZ_I` which has relatively unoptimised arithmetic (that is something that I've found in quite a few benchmarks).

oscarbenjamin commented:
The matrices from gh-25403 can be computed in about 10 seconds with the PR using `DomainMatrix` but are still slower when using `Matrix`:
```python
In [5]: %time ok = M1.to_DM().charpoly()
CPU times: user 10.9 s, sys: 3.97 ms, total: 10.9 s
Wall time: 10.9 s
```
The reason `Matrix.charpoly` is slower than DomainMatrix is this code that I did not immediately want to change:
https://github.com/sympy/sympy/blob/0f705ba4587ce7a55ba6f153bb802fb2c61a574c/sympy/matrices/determinant.py#L416-L425
The slow part here is actually converting back to Expr (`to_sympy`). I left that there so that the `simplify` function could still be used. The Matrix.charpoly method allows a user to supply the simplify function like:
```python
In [6]: M = Matrix([[1/(sqrt(2) + sqrt(3))]])

In [7]: M.charpoly()
Out[7]: PurePoly(lambda - 1/(sqrt(2) + sqrt(3)), lambda, domain='EX')

In [8]: M.charpoly(simplify=radsimp)
Out[8]: PurePoly(lambda - sqrt(3) + sqrt(2), lambda, domain='EX')
```
We can't support that in a fully backwards compatible way without converting back to `Expr`. It would be much more efficient to convert the characteristic polynomial `cp` directly to `Poly` like: `Poly(cp, x, domain=dM.domain)`. Perhaps this can just check if the user has supplied the simplify function and avoid converting back to Expr if so. We could call `uniquely_named_symbol` on the original matrix elements rather than the charpoly coefficients.

oscarbenjamin committed:
perf(matrices): avoid Expr conversion in charpoly

oscarbenjamin commented:
> We can't support that in a fully backwards compatible way without converting back to `Expr`. It would be much more efficient to convert the characteristic polynomial `cp` directly to `Poly` like: `Poly(cp, x, domain=dM.domain)`. Perhaps this can just check if the user has supplied the simplify function and avoid converting back to Expr if so. We could call `uniquely_named_symbol` on the original matrix elements rather than the charpoly coefficients.

I've just pushed a commit that does this. Now the matrices from https://github.com/sympy/sympy/issues/25403 can be computed in 10 seconds each:
```python
In [2]: %time ok = M1.charpoly()
CPU times: user 10.1 s, sys: 19.7 ms, total: 10.1 s
Wall time: 10.1 s

In [3]: %time ok = M2.charpoly()
CPU times: user 10.5 s, sys: 4.45 ms, total: 10.5 s
Wall time: 10.5 s
```
There is plenty that can be improved there but still the charpoly code is doing its job. Note that the output is enormous:
```python
In [4]: len(str(ok))
Out[4]: 1172032
```
In this example we can actually compute factors of the characteristic polynomial more efficiently:
```python
In [7]: %time ok = M2.to_DM().charpoly_factor_blocks()
CPU times: user 87.5 ms, sys: 4.04 ms, total: 91.5 ms
Wall time: 90.5 ms
```
That's the `DomainMatrix` code doing its job efficiently. The slow part is that for `charpoly` we have to multiply those factors together and that takes 10 seconds. Alternatively for `charpoly_factor_list` we need to factorise each factor and that is very slow. Those are separate issues from the code changed in this PR though which does its part in reasonable time.

oscarbenjamin commented:
> If the domain ends up being `EX` then this might not be faster than master e.g. this happens in about the same time for both master and the PR:

Actually I was wrong about this. The reason some examples had about the same time with the PR and with master is because all the time was spent in the call to `simplify` and `uniquely_named_symbol` which I have now removed. Some examples from above are much faster since the last commit 
5c79dab e.g.:
```python
In [2]: M = randMatrix(5)*sin(x) + randMatrix(5)*cos(x)

In [3]: %time ok = M.charpoly()
CPU times: user 233 ms, sys: 250 µs, total: 233 ms
Wall time: 232 ms
```
That took 13 seconds before because of the call to `simplify`. It is probably not producing as simple output though because this particular example with cos and sin is something that simplify would be able to make simpler:
```python
In [10]: M = randMatrix(3)*sin(x) + randMatrix(3)*cos(x)

In [11]: M.charpoly()
Out[11]: 
PurePoly(lambda**3 + (-80*sin(x) - 140*cos(x))*lambda**2 + ((-43*sin(x) - 122*cos(x))*(-37*sin(x) -
 18*cos(x)) + (-25*sin(x) - 93*cos(x))*(-18*sin(x) - 29*cos(x)) - (21*sin(x) + 7*cos(x))*(65*sin(x)
 + 64*cos(x)) - (37*sin(x) + 37*cos(x))*(88*sin(x) + 35*cos(x)) - (78*sin(x) + 71*cos(x))*(88*sin(x
) + 69*cos(x)))*lambda + ((-25*sin(x) - 93*cos(x))*(-18*sin(x) - 29*cos(x)) - (78*sin(x) + 71*cos(x
))*(88*sin(x) + 69*cos(x)))*(-37*sin(x) - 18*cos(x)) - ((18*sin(x) + 29*cos(x))*(21*sin(x) + 7*cos(
x)) + (78*sin(x) + 71*cos(x))*(88*sin(x) + 35*cos(x)))*(65*sin(x) + 64*cos(x)) + (-(21*sin(x) + 7*c
os(x))*(65*sin(x) + 64*cos(x)) - (37*sin(x) + 37*cos(x))*(88*sin(x) + 35*cos(x)))*(-43*sin(x) - 122
*cos(x)) - ((21*sin(x) + 7*cos(x))*(88*sin(x) + 69*cos(x)) + (25*sin(x) + 93*cos(x))*(88*sin(x) + 3
5*cos(x)))*(37*sin(x) + 37*cos(x)), lambda, domain='EXRAW')

In [12]: print(M.charpoly().as_expr().simplify())
lambda**3 - 80*lambda**2*sin(x) - 140*lambda**2*cos(x) - 10293*lambda*sin(2*x)/2 + 7695*lambda*cos(2*x)/2 - 11193*lambda/2 - 200605*sin(x) - 16120*sin(3*x) - 268277*cos(x)/2 + 152165*cos(3*x)/2
```
Maybe then if the domain was `EX` then the call to `simplify` should be restored.

oscarbenjamin commented:
This example is from an issue somewhere (I don't remember where right now):
```python
In [1]: a, b, c, d, e = symbols('a:e')
   ...: 
   ...: M = Matrix([
   ...: [      0,    1,       0,    0,       0,    0,    0],
   ...: [-1/a**2, -2/a,       0,    0,       0,    0,    0],
   ...: [      0,    0,       0,    1,       0,    0,    0],
   ...: [      0,    0, -1/b**2, -2/b,       0,    0,    0],
   ...: [      0,    0,       0,    0,       0,    1,    0],
   ...: [      0,    0,       0,    0, -1/c**2, -2/c,    0],
   ...: [    1/d,    0,     1/d,    0,     1/d,    0, -1/e]])
   ...: 

In [2]: %time ok = M.charpoly()  # time with master
CPU times: user 3.62 s, sys: 14.8 ms, total: 3.63 s
Wall time: 3.63 s
```
With the PR it is a lot faster:
```python
In [1]: a, b, c, d, e = symbols('a:e')
   ...: 
   ...: M = Matrix([
   ...: [      0,    1,       0,    0,       0,    0,    0],
   ...: [-1/a**2, -2/a,       0,    0,       0,    0,    0],
   ...: [      0,    0,       0,    1,       0,    0,    0],
   ...: [      0,    0, -1/b**2, -2/b,       0,    0,    0],
   ...: [      0,    0,       0,    0,       0,    1,    0],
   ...: [      0,    0,       0,    0, -1/c**2, -2/c,    0],
   ...: [    1/d,    0,     1/d,    0,     1/d,    0, -1/e]])

In [2]: %time ok = M.charpoly()
CPU times: user 26.9 ms, sys: 3.83 ms, total: 30.8 ms
Wall time: 31.1 ms
```
So that's about 100x faster with the PR. This is a good example for the benefit of clearing denominators.

oscarbenjamin committed:
fix(matrices): simplify charpoly if domain is EX

oscarbenjamin commented:
> Maybe then if the domain was `EX` then the call to `simplify` should be restored.

I have restored this in the last commit. Now `simplify` is used if the domain was `EX` or if a custom simplify function was provided to `charpoly`.

There is an interesting test failure here:
https://github.com/sympy/sympy/blob/fc7228549fb9d752865d3ec4739d69a26e4fc0d9/sympy/matrices/tests/test_eigen.py#L145-L154
The test asserts that calling eigenvals with `simplify=True` will reduce the op count but it does not:
```python
In [14]: q = symbols('q', positive=True)

In [15]: M = Matrix([
    ...: [    -2, exp(-q),  1],
    ...: [exp(q),      -2,  1],
    ...: [     1,       1, -2]])

In [16]: print(repr(M.eigenvals()))
{-(sqrt((27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q) - 54)**2 - 2916)/2 + 27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q)/2 - 27)**(1/3)/3 - 2 - 3/(sqrt((27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q) - 54)**2 - 2916)/2 + 27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q)/2 - 27)**(1/3): 1, -(-1/2 + sqrt(3)*I/2)*(sqrt((27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q) - 54)**2 - 2916)/2 + 27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q)/2 - 27)**(1/3)/3 - 2 - 3/((-1/2 + sqrt(3)*I/2)*(sqrt((27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q) - 54)**2 - 2916)/2 + 27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q)/2 - 27)**(1/3)): 1, -(-1/2 - sqrt(3)*I/2)*(sqrt((27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q) - 54)**2 - 2916)/2 + 27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q)/2 - 27)**(1/3)/3 - 2 - 3/((-1/2 - sqrt(3)*I/2)*(sqrt((27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q) - 54)**2 - 2916)/2 + 27*(-exp(2*q) + 2*exp(q) - 1)*exp(-q)/2 - 27)**(1/3)): 1}

In [17]: print(repr(M.eigenvals(simplify=True)))
{-2 - (2*sqrt(exp(4*q) - 2*exp(2*q) + 1) - 2*exp(2*q) - 2)**(2/3)*exp(-q/3)/(2*(sqrt(exp(4*q) - 2*exp(2*q) + 1) - exp(2*q) - 1)**(1/3)) - 2**(1/3)*exp(q/3)/(sqrt(exp(4*q) - 2*exp(2*q) + 1) - exp(2*q) - 1)**(1/3): 1, (2**(2/3)*(1 - sqrt(3)*I)**2*(sqrt((exp(2*q) + 1)**2 - 4*exp(2*q)) - exp(2*q) - 1)**(2/3) + 8*(-1 + sqrt(3)*I)*(sqrt((exp(2*q) + 1)**2 - 4*exp(2*q)) - exp(2*q) - 1)**(1/3)*exp(q/3) + 8*2**(1/3)*exp(2*q/3))*exp(-q/3)/(4*(1 - sqrt(3)*I)*(sqrt((exp(2*q) + 1)**2 - 4*exp(2*q)) - exp(2*q) - 1)**(1/3)): 1, (2**(2/3)*(1 + sqrt(3)*I)**2*(sqrt((exp(2*q) + 1)**2 - 4*exp(2*q)) - exp(2*q) - 1)**(2/3) - 8*(1 + sqrt(3)*I)*(sqrt((exp(2*q) + 1)**2 - 4*exp(2*q)) - exp(2*q) - 1)**(1/3)*exp(q/3) + 8*2**(1/3)*exp(2*q/3))*exp(-q/3)/(4*(1 + sqrt(3)*I)*(sqrt((exp(2*q) + 1)**2 - 4*exp(2*q)) - exp(2*q) - 1)**(1/3)): 1}

In [18]: count_ops(M.eigenvals(simplify=False))
Out[18]: 219

In [19]: count_ops(M.eigenvals(simplify=True))
Out[19]: 231
```
I don't see the purpose of that test. Here it is the `simplify` function that increases the op-count but that is not what is being tested. Also op-count is not a good measure of simplicity anyway.

On master both outputs are simpler:
```python
In [1]: q = symbols('q', positive=True)

In [2]: M = Matrix([
   ...: [    -2, exp(-q),  1],
   ...: [exp(q),      -2,  1],
   ...: [     1,       1, -2]])

In [3]: print(repr(M.eigenvals(simplify=True)))
{2*cosh(q/3) - 2: 1, (sqrt(3)*I*(exp(q/3) + 2) + exp(q/3) - 2 - 2*exp(-q/3))/(1 - sqrt(3)*I): 1, (-sqrt(3)*I*(exp(q/3) + 2) + exp(q/3) - 2 - 2*exp(-q/3))/(1 + sqrt(3)*I): 1}

In [4]: print(repr(M.eigenvals(simplify=False)))
{(sqrt(cosh(q)**2 - 1) + cosh(q))**(1/3) - 2 + (sqrt(cosh(q)**2 - 1) + cosh(q))**(-1/3): 1, (-1/2 + sqrt(3)*I/2)*(sqrt(cosh(q)**2 - 1) + cosh(q))**(1/3) - 2 + 1/((-1/2 + sqrt(3)*I/2)*(sqrt(cosh(q)**2 - 1) + cosh(q))**(1/3)): 1, (-1/2 - sqrt(3)*I/2)*(sqrt(cosh(q)**2 - 1) + cosh(q))**(1/3) - 2 + 1/((-1/2 - sqrt(3)*I/2)*(sqrt(cosh(q)**2 - 1) + cosh(q))**(1/3)): 1}
```
The real question though is what happened to the charpoly. With master we get:
```python
In [5]: M.charpoly()
Out[5]: PurePoly(lambda**3 + 6*lambda**2 + 9*lambda + 2 - 2*cosh(q), lambda, domain='ZZ[cosh(q)]')
```
With the PR we get:
```
In [3]: M.charpoly()
Out[3]: 
PurePoly(lambda**3 + 6*lambda**2 + 9*lambda + (-exp(2*q) + 2*exp(q) - 1)*exp(-q), lambda, domain='ZZ(ex
p(q))')
```
So simplify has turned the exponentials into `cosh` on master. I've now disabled the use of simplify in the PR so we get a result in terms of exponentials and that fails the test.

I don't think that it should be the job of `charpoly` to turn `exp` into `cosh` so I think that the test should just be changed.

oscarbenjamin commented:
> I don't think that it should be the job of `charpoly` to turn `exp` into `cosh` so I think that the test should just be changed.

@sylee957 this test comes from gh-15125. Do you have an opinion about this?

sylee957 commented:
I think that the test can use the variance
`m.eigenvals(simplify=False) != m.eigenvals(simplify=True)`
if simplify doesn't always reduce the `count_ops`.
The tests shouldn't have much semantics than snapshots.

oscarbenjamin commented:
> I think that the test can use the variance
> `m.eigenvals(simplify=False) != m.eigenvals(simplify=True)`

What value would testing that provide?

I don't see how it would be a "failure" if the same result was returned in both cases but that would show as a fail if the test used `!=`. If eigenvals happened to return a simplified result by default without simplify=True then that would be fine and if simplify didn't simplify that result then that might also be fine if the result was already as simple as possible. In fact *ideally* this is what would happen: `eigenvals` would already return the simplest possible result in all cases. This test would reject that...

sylee957 commented:
I'm fine to either remove the test. 
however, ideally, the test should be snapshot testing of expressions,
but I didn't want to do that exactly because writing down long expressions in the test cases are awful,
and the only way to do this is approximating that tests, even if that doesn't reflect how `simplify` is supposed to work or not.
For example, if someone attempts to change `simplify` itself, some of the test cases should reflect what kind of examples it changes.

oscarbenjamin committed:
fix(matrices): restore the use of EXRAW iin charpoly

Using EX caused test_second_order_to_first_order_slow4 to fail.

A better fix would be to change something else but for now we go back to
using EXRAW instead of EX just to match previous behaviour.

oscarbenjamin commented:
This test is now failing in `tests-latest (4)`:
https://github.com/sympy/sympy/blob/fc7228549fb9d752865d3ec4739d69a26e4fc0d9/sympy/solvers/ode/tests/test_systems.py#L1740-L1755

It hangs in `integrate` because the form of the expression passed to `integrate` has changed. In this case `dsolve` actually should not call `integrate` at all. The test passes `doit=False` which is supposed to prevent integration and there is a comment noting that `integrate` hangs although it apparently does not with master.

The cause of the difference is this. With master we have:
```python
In [17]: M.charpoly()
Out[17]: PurePoly(lambda**2 - (t*cos(t) - sin(t))**2, lambda, domain='EX')

In [18]: M.eigenvals()
Out[18]: {-t⋅cos(t) + sin(t): 1, t⋅cos(t) - sin(t): 1}
```
With the PR we have:
```python
In [1]: M = Matrix([
   ...: [                 0, -t*cos(t) + sin(t)],
   ...: [-t*cos(t) + sin(t),                  0]])

In [2]: M.charpoly()
Out[2]: PurePoly(lambda**2 - t**2*cos(t)**2 + t*sin(2*t) - sin(t)**2, lambda, domain='EX')

In [3]: M.eigenvals()
Out[3]: 
⎧    ___________________________________        ___________________________________   ⎫
⎨   ╱  2    2                      2           ╱  2    2                      2       ⎬
⎩-╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1, ╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1⎭
```
That difference is because I switch to use `EX` rather than `EXRAW` which expands the power. It should be possible for `roots` to factorise the power e.g.:
```python
In [5]: roots(z**2 - (x - y)**2, z)
Out[5]: {-x + y: 1, x - y: 1}

In [6]: roots(z**2 - expand((x - y)**2), z)
Out[6]: {-x + y: 1, x - y: 1}
```
I have preserved the `charpoly` behaviour of using `simplify` if the domain is `EX` though and `simplify` turns `sin(t)*cos(t)` into `sin(2*t)` which then means it cannot be factorised.

I think that really this is something that should be handled in `roots` rather than `charpoly`. Probably `roots_binomial` should use `factor` if possible. The output from `charpoly` is not ideal because `simplify` makes the expression worse by changing `sin(t)*cos(t)` to `sin(2*t)` which prevents factorisation.

I think it is fine that charpoly outputs a result with an expanded power. If anything the problem is that `charpoly` should not use `simplify` at all. I have tried to reduce the use of `simplify` here but it is difficult to change that without breaking some sort of compatibility somewhere.

I think that for now I will restore the use of `EXRAW` instead of `EX`. We can let `simplify` do all of the simplifying as it currently does with master. That is a lot slower when a proper domain is not found but it preserves existing behaviour on master for that case.

oscarbenjamin commented:
> I think that for now I will restore the use of `EXRAW` instead of `EX`. We can let `simplify` do all of the simplifying as it currently does with master. That is a lot slower when a proper domain is not found but it preserves existing behaviour on master for that case.

I have just pushed this and so that failing test passes and existing behaviour is preserved for when the domain is EX:
```python
In [1]: M = Matrix([
   ...: [                 0, -t*cos(t) + sin(t)],
   ...: [-t*cos(t) + sin(t),                  0]])

In [2]: M.eigenvals()
Out[2]: {-t⋅cos(t) + sin(t): 1, t⋅cos(t) - sin(t): 1}

In [3]: M.charpoly()
Out[3]: PurePoly(lambda**2 - (-t*cos(t) + sin(t))**2, lambda, domain='EX')
```

oscarbenjamin committed:
fix(matrices): fix is_EXRAW check

oscarbenjamin committed:
test(matrices): remove eigenvals simplify test

oscarbenjamin commented on sympy/matrices/tests/test_eigen.py:
```diff
@@ -142,18 +142,6 @@ def NS(e, n):
     assert isinstance(m.eigenvals(simplify=lambda x: x, multiple=True), list)
 
 
-@slow
-def test_eigen_slow():
-    # issue 15125
-    from sympy.core.function import count_ops
-    q = Symbol("q", positive = True)
-    m = Matrix([[-2, exp(-q), 1], [exp(q), -2, 1], [1, 1, -2]])
-    assert count_ops(m.eigenvals(simplify=False)) > \
-        count_ops(m.eigenvals(simplify=True))
-    assert count_ops(m.eigenvals(simplify=lambda x: x)) > \
-        count_ops(m.eigenvals(simplify=True))
```
I'm removing this test as discussed above. The test fails here because simplify increases the op count but that is not really a useful test because it is really testing simplify rather than eigenvals and also it is not necessarily a bug in simplify (or eigenvals) if the op count increases.

There could perhaps be a better way to test for the effect of the simplify parameter but really I would prefer it if methods like `eigenvals` did not use generic `simplify` at all. The test passes on master because simplify changes `exp` into `cosh` which is considered to reduce the op count but I don't think that changing exp into cosh is something that either eigenvals or charpoly should do. It might seem that cosh is simpler than exp in the context of some particular expression but e.g. if you have other expressions involving exp then it is better to keep them all in terms of exp rather than change some to cosh.

oscarbenjamin committed:
matrices: remove redundant import

oscarbenjamin commented:
This is ready for review.

One open question is whether to use `EX` rather than `EXRAW` if a more suitable domain cannot be found. Currently the PR uses EXRAW. This is slow because EXRAW generates large unsimplified output and then `simplify` is used to simplify that down again. Using `EX` means that the input to `simplify` is computed a bit more slowly but is already simpler so `simplify` is faster which means that the total time is significantly less. Using `EXRAW+simplify` essentially corresponds to the current method on master and is the main reason that `charpoly` is slow on maser.

Changing to use `EX` instead of `EXRAW` would make many cases a lot faster e.g. `(randMatrix(5) + randMatrix(5)*sqrt(2)).charpoly()` would be 1.4 seconds instead of 30 seconds so that's 20x faster but the ratio increases if you make the matrices larger e.g. for `n=6` it is 180x faster and so on.

Specific examples like this `sqrt(2)` matrix can be handled much more efficiently in other ways anyway e.g. `(randMatrix(5) + randMatrix(5)*sqrt(2)).to_DM(extension=True).charpoly()` is more like 20 milliseconds so 1000x faster than EXRAW or 70x faster than EX and again the difference would increase as you scale up. The proper fix to make that work is a bit more complicated than just using `extension=True` so I haven't implemented it here yet. In general though there will always be some residual cases that the domain system cannot completely handle and so the question is whether to use `EX` for those cases which is a lot faster than EXRAW according to any comparisons I have done.

The downside of using `EX` rather than `EXRAW` is that this example does not come out as nice:
```python
In [1]: M = Matrix([
   ...: [                 0, -t*cos(t) + sin(t)],
   ...: [-t*cos(t) + sin(t),                  0]])

In [2]: M.charpoly()
Out[2]: PurePoly(lambda**2 - (t*cos(t) - sin(t))**2, lambda, domain='EX')

In [3]: M.eigenvals()
Out[3]: {-t⋅cos(t) + sin(t): 1, t⋅cos(t) - sin(t): 1}
```
Instead with `EX` it would be:
```python
In [1]: M = Matrix([
   ...: [                 0, -t*cos(t) + sin(t)],
   ...: [-t*cos(t) + sin(t),                  0]])

In [2]: M.charpoly()
Out[2]: PurePoly(lambda**2 - t**2*cos(t)**2 + t*sin(2*t) - sin(t)**2, lambda, domain='EX')

In [3]: M.eigenvals()
Out[3]: 
⎧    ___________________________________        ___________________________________   ⎫
⎨   ╱  2    2                      2           ╱  2    2                      2       ⎬
⎩-╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1, ╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1⎭
```
This is also an example of something that can eventually be improved by having a proper domain that can represent sin and cos. I think really this is something that should be fixed elsewhere. The new charpoly output is closer to what should be the expected expanded canonical output for charpoly. What is wrong with it is that charpoly uses `simplify` and `simplify` changes the expression but not in a good way. Ideally `simplify` would factor it or leave it alone and then `roots` should factor it to give simple expressions for the roots.

This matrix `M` comes up in `test_second_order_to_first_order_slow4` which would have to marked as XFAIL if charpoly was changed to use EX instead of EXRAW. The changed expression causes the test to hang in integrate. Also a bug in dsolve should be fixed because actually the test is not supposed to be use `integrate` at all. I don't want to change that here though.

My suggestion here is that charpoly should use `EX` and `test_second_order_to_first_order_slow4` should be marked as XFAIL but conservatively I have not done that yet in this PR. Basically the current output using `EXRAW` more closely resembles the existing output on master for those cases so I left it at that for now.

Another question to consider is if the call to `simplify` in `charpoly` can be removed if `EX` is being used. Basically `EX` already gives the sort of simplification that is typically wanted here (basically `expand`) so the only reason to use general `simplify` is for it to do more obscure kinds of simplification for example in the test that was removed we have something like:
```python
In [4]: simplify(x**2 + x + exp(t) + exp(-t))
Out[4]: 
 2                
x  + x + 2⋅cosh(t)
```
I don't think that it should be the job of charpoly to do that sort of simplification though: the caller can call `simplify` on the output if they want. Currently the PR uses `EXRAW` and then calls `simplify` but I think it should be possible to change to using `EX` and then not calling `simplify`. Not using `simplify` would help with the matrix `M` above since then the trig part could then be factorised by `roots`.

For now using `EXRAW+simplify` is the more conservative change relative to master so perhaps it would be better to merge this as is and then I could open a separate PR to make that as a followup change.

smichr commented:
> so I think that the test should just be changed.

I suspect the test was just testing that `simplify` did something.

OI added the reference to the issue for the charpoly example you gave.

smichr commented:
> I don't think that it should be the job of charpoly to do that sort of simplification though: the caller can call simplify on the output if they want. 

I don't mind moving away from many functions automatically simplifying output toward less doing so and leaving the decision to simplify up to the user.

smichr commented on sympy/polys/matrices/sdm.py:
```diff
@@ -1243,6 +1268,22 @@ def sdm_transpose(M):
 
 
+def sdm_dotvec(A, B, K):
+    total = K.zero
+    for j in A.keys() & B.keys():
+        total += A[j] * B[j]
+    return total
```
```suggestion
    return sum(A[j]*B[j] for j in A.keys() & B.keys(), start=K.zero)
```

oscarbenjamin replied:
Yes, good point. Actually `K.sum` is better (it uses `Add(*args)` for `EXRAW`).

smichr commented on sympy/polys/matrices/sdm.py:
```diff
@@ -1831,3 +1872,126 @@ def sdm_particular_from_rref(A, ncols, pivots):
+        return {0: one}
+    elif n == 1:
+        pdict = {0: one}
+        M0 = M.get(0)
+        if M0 is not None:
+            M00 = M0.get(0, zero)
+            if M00:
+                pdict[1] = -M00
```
```suggestion
        if M00 := M.get(0, {}).get(0, zero)
            pdict[1] = -M00
```

oscarbenjamin replied:
Yes, fair enough. We probably don't need to optimise too much for `n=1`.

smichr commented on sympy/polys/matrices/sdm.py:
```suggestion
        if Tqi := sdm_dotvec(Ti, q, K):
            Tq[i] = Tqi
```

smichr commented on sympy/polys/matrices/sdm.py:
```suggestion
    for i in range(min(q), min(max(q)+len(Tvals), n+1)):
```

oscarbenjamin committed:
matrices: code tweaks in DomainMatrix

oscarbenjamin commented:
Thanks @smichr. I've pushed those changes.

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_charpoly

oscarbenjamin commented:
Looks like the python-flint job fails here after gh-25495.

The `nnz` method needs to be added to `DFM` now as well in this PR.

oscarbenjamin committed:
fix(matrices): add DFM.nnz method

oscarbenjamin commented:
Thanks @smichr!
